pinot-core/src/main/java/org/apache/pinot/core/data/table/IndexedTableResizer.java

```diff
@@ -0,0 +1,307 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data.table;
+
+import com.google.common.annotations.VisibleForTesting;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.function.Function;
+import org.apache.pinot.common.request.AggregationInfo;
+import org.apache.pinot.common.request.SelectionSort;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunctionUtils;
+
+
+/**
+ * Helper class for trimming and sorting records in the IndexedTable, based on the order by information
+ */
+class IndexedTableResizer {
+
+  private OrderByValueExtractor[] _orderByValueExtractors;
+  private Comparator<IntermediateRecord> _intermediateRecordComparator;
+  private int _numOrderBy;
+
+  IndexedTableResizer(DataSchema dataSchema, List<AggregationInfo> aggregationInfos, List<SelectionSort> orderBy) {
+
+    int numAggregations = aggregationInfos.size();
+    int numKeyColumns = dataSchema.size() - numAggregations;
+
+    Map<String, Integer> keyIndexMap = new HashMap<>();
+    for (int i = 0; i < numKeyColumns; i++) {
+      String columnName = dataSchema.getColumnName(i);
+      keyIndexMap.put(columnName, i);
+    }
+
+    Map<String, Integer> aggregationColumnToIndex = new HashMap<>();
```

Comment:
Store key index and aggregation column index into the same map, use the index to determine whether it is key or aggregation column

Reply:
The set of indexes for keys and columns is not a disjoint set. For example, 
`select sum(a), sum(b) from table group by c order by c`
the keyToIndex map is `{c -> 0}` and aggregationToIndex map is `{sum(a) -> 0, sum(b) -> 1}`
This is because the key[] and value[] are stored as separate arrays in the Table.
So I cannot use the index to determine the column type right?

Reply:
after the offline discussion, changed it to 1 map